File: ma_sh/Model/mash.py
import torch
from typing import Tuple

import mash_cpp

from ma_sh.Model.base_mash import BaseMash
from ma_sh.Method.Mash.mash import toPreLoadDatas
from ma_sh.Method.rotate import toRotateVectorsFromOrthoPoses


class Mash(BaseMash):
    def __init__(
        self,
        anchor_num: int = 400,
        mask_degree_max: int = 3,
        sh_degree_max: int = 2,
        use_inv: bool = True,
        dtype=torch.float64,
        device: str = "cuda" if torch.cuda.is_available() else "cpu",
        mask_boundary_sample_num: int = 90,
        sample_polar_num: int = 1000,
        sample_point_scale: float = 0.8,
    ) -> None:
        # Super Params
        self.mask_boundary_sample_num = mask_boundary_sample_num
        self.sample_polar_num = sample_polar_num
        self.sample_point_scale = sample_point_scale

        super().__init__(
            anchor_num,
            mask_degree_max,
            sh_degree_max,
            use_inv,
            dtype,
            device,
        )
        return

    def updatePreLoadDatas(self) -> bool:
        (
            self.sample_phis,
            self.sample_thetas,
            self.mask_boundary_phis,
            self.mask_boundary_phi_idxs,
            self.mask_boundary_base_values,
            self.sample_base_values,
        ) = toPreLoadDatas(
            self.anchor_num,
            self.mask_degree_max,
            self.mask_boundary_sample_num,
            self.sample_polar_num,
            self.dtype,
            self.device,
        )
        return True

    def toMaskBoundaryThetas(self) -> torch.Tensor:
        mask_boundary_thetas = mash_cpp.toMaskBoundaryThetas(
            self.mask_params,
            self.mask_boundary_base_values,
            self.mask_boundary_phi_idxs,
        )
        return mask_boundary_thetas

    def toInMaskSamplePolars(
        self, mask_boundary_thetas: torch.Tensor
    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
        (
            in_mask_sample_phis,
            in_mask_sample_theta_weights,
            in_mask_sample_polar_idxs,
            in_mask_sample_base_values,
        ) = mash_cpp.toInMaskSamplePolars(
            self.anchor_num,
            self.mask_params,
            self.sample_phis,
            self.sample_thetas,
            mask_boundary_thetas,
            self.mask_boundary_phi_idxs,
            self.sample_base_values,
        )

        return (
            in_mask_sample_phis,
            in_mask_sample_theta_weights,
            in_mask_sample_polar_idxs,
            in_mask_sample_base_values,
        )

    def toSamplePointsTuple(
        self,
    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
        rotate_vectors = toRotateVectorsFromOrthoPoses(self.ortho_poses)

        (
            mask_boundary_sample_points,
            in_mask_sample_points,
            in_mask_sample_point_idxs,
        ) = mash_cpp.toMashSamplePoints(
            self.anchor_num,
            self.mask_degree_max,
            self.sh_degree_max,
            self.mask_params,
            self.sh_params,
            rotate_vectors,
            self.positions,
            self.sample_phis,
            self.sample_thetas,
            self.mask_boundary_phis,
            self.mask_boundary_phi_idxs,
            self.mask_boundary_base_values,
            self.sample_base_values,
            self.sample_point_scale,
            self.use_inv,
        )

        return (
            mask_boundary_sample_points,
            in_mask_sample_points,
            in_mask_sample_point_idxs,
        )

    def toSamplePoints(self) -> torch.Tensor:
        (
            mask_boundary_sample_points,
            in_mask_sample_points,
        ) = self.toSamplePointsTuple()[:2]

        sample_points = torch.vstack(
            [mask_boundary_sample_points, in_mask_sample_points]
        )

        return sample_points

File: ma_sh/Cpp/src/mash.cpp
#include "mash.h"
#include "constant.h"
#include "direction.h"
#include "fps.h"
#include "mash_unit.h"
#include "mask.h"
#include "timer.h"

const std::vector<torch::Tensor>
toInMaskSamplePolars(const int &anchor_num, const torch::Tensor &mask_params,
                     const torch::Tensor &sample_phis,
                     const torch::Tensor &sample_thetas,
                     const torch::Tensor &mask_boundary_thetas,
                     const torch::Tensor &mask_boundary_phi_idxs,
                     const torch::Tensor &sample_base_values) {
  const std::vector<torch::Tensor> in_max_mask_sample_polar_idxs_vec =
      toInMaxMaskSamplePolarIdxsVec(anchor_num, sample_thetas,
                                    mask_boundary_thetas,
                                    mask_boundary_phi_idxs);

  const torch::Tensor in_max_mask_sample_polar_idxs =
      toInMaxMaskSamplePolarIdxs(in_max_mask_sample_polar_idxs_vec);

  const torch::Tensor in_max_mask_sample_polar_data_idxs =
      torch::hstack(in_max_mask_sample_polar_idxs_vec);

  const torch::Tensor in_max_mask_sample_phis =
      sample_phis.index({in_max_mask_sample_polar_data_idxs});

  const torch::Tensor in_max_mask_sample_thetas =
      sample_thetas.index({in_max_mask_sample_polar_data_idxs});

  const torch::Tensor in_max_mask_sample_base_values =
      sample_base_values.index({slice_all, in_max_mask_sample_polar_data_idxs});

  const torch::Tensor in_max_mask_thetas = toInMaxMaskThetas(
      mask_params, in_max_mask_sample_base_values,
      in_max_mask_sample_polar_idxs, in_max_mask_sample_polar_data_idxs);

  const torch::Tensor in_mask_sample_polar_mask =
      in_max_mask_sample_thetas <= in_max_mask_thetas;

  const torch::Tensor in_mask_sample_phis =
      in_max_mask_sample_phis.index({in_mask_sample_polar_mask});

  const torch::Tensor in_mask_sample_polar_idxs =
      in_max_mask_sample_polar_idxs.index({in_mask_sample_polar_mask});

  const torch::Tensor in_mask_sample_polar_data_idxs =
      in_max_mask_sample_polar_data_idxs.index({in_mask_sample_polar_mask});

  const torch::Tensor in_mask_sample_base_values =
      in_max_mask_sample_base_values.index(
          {slice_all, in_mask_sample_polar_mask});

  const torch::Tensor in_mask_sample_theta_weights = toInMaskSampleThetaWeights(
      in_max_mask_sample_thetas, in_max_mask_thetas, in_mask_sample_polar_mask);

  const std::vector<torch::Tensor> in_mask_sample_polars_with_idxs(
      {in_mask_sample_phis, in_mask_sample_theta_weights,
       in_mask_sample_polar_idxs, in_mask_sample_base_values});
  return in_mask_sample_polars_with_idxs;
}

const torch::Tensor
toSamplePoints(const int &mask_degree_max, const int &sh_degree_max,
               const torch::Tensor &sh_params,
               const torch::Tensor &rotate_vectors,
               const torch::Tensor &positions, const torch::Tensor &sample_phis,
               const torch::Tensor &sample_thetas,
               const torch::Tensor &sample_polar_idxs, const bool &use_inv) {
  const torch::Tensor sh_values = toSHValues(
      sh_degree_max, sh_params, sample_phis, sample_thetas, sample_polar_idxs);

  const torch::Tensor sample_sh_directions =
      toDirections(sample_phis, sample_thetas);

  const torch::Tensor sh_points =
      toSHPoints(sh_params, rotate_vectors, positions, sample_sh_directions,
                 sh_values, sample_polar_idxs, use_inv);

  return sh_points;
}

const torch::Tensor toWeightedSamplePoints(
    const int &mask_degree_max, const int &sh_degree_max,
    const torch::Tensor &mask_params, const torch::Tensor &sh_params,
    const torch::Tensor &rotate_vectors, const torch::Tensor &positions,
    const torch::Tensor &sample_phis, const torch::Tensor &sample_theta_weights,
    const torch::Tensor &sample_polar_idxs, const bool &use_inv,
    const torch::Tensor &sample_base_values) {
  torch::Tensor valid_sample_base_values = sample_base_values;
  if (sample_base_values.size(0) != sample_phis.size(0)) {
    valid_sample_base_values = toMaskBaseValues(sample_phis, mask_degree_max);
  }

  const torch::Tensor detect_thetas =
      toDetectThetas(mask_params, valid_sample_base_values, sample_polar_idxs,
                     sample_theta_weights);

  const torch::Tensor sh_points = toSamplePoints(
      mask_degree_max, sh_degree_max, sh_params, rotate_vectors, positions,
      sample_phis, detect_thetas, sample_polar_idxs, use_inv);

  return sh_points;
}

const std::vector<torch::Tensor> toMashSamplePoints(
    const int &anchor_num, const int &mask_degree_max, const int &sh_degree_max,
    const torch::Tensor &mask_params, const torch::Tensor &sh_params,
    const torch::Tensor &rotate_vectors, const torch::Tensor &positions,
    const torch::Tensor &sample_phis, const torch::Tensor &sample_thetas,
    const torch::Tensor &mask_boundary_phis,
    const torch::Tensor &mask_boundary_phi_idxs,
    const torch::Tensor &mask_boundary_base_values,
    const torch::Tensor &sample_base_values, const float &sample_point_scale,
    const bool &use_inv) {
  Timer timer;
  const torch::Tensor mask_boundary_thetas = toMaskBoundaryThetas(
      mask_params, mask_boundary_base_values, mask_boundary_phi_idxs);

  timer.reset();
  const std::vector<torch::Tensor> in_mask_sample_polars_with_idxs =
      toInMaskSamplePolars(anchor_num, mask_params, sample_phis, sample_thetas,
                           mask_boundary_thetas, mask_boundary_phi_idxs,
                           sample_base_values);
  std::cout << "toInMaskSamplePolars : " << timer.now() << std::endl;

  const torch::Tensor &in_mask_sample_phis = in_mask_sample_polars_with_idxs[0];
  const torch::Tensor &in_mask_sample_theta_weights =
      in_mask_sample_polars_with_idxs[1];
  const torch::Tensor &in_mask_sample_polar_idxs =
      in_mask_sample_polars_with_idxs[2];
  const torch::Tensor &in_mask_sample_base_values =
      in_mask_sample_polars_with_idxs[3];

  timer.reset();
  const torch::Tensor in_mask_sh_points = toWeightedSamplePoints(
      mask_degree_max, sh_degree_max, mask_params, sh_params, rotate_vectors,
      positions, in_mask_sample_phis, in_mask_sample_theta_weights,
      in_mask_sample_polar_idxs, use_inv, in_mask_sample_base_values);
  std::cout << "toWeightedSamplePoints : " << timer.now() << std::endl;

  timer.reset();
  const torch::Tensor fps_in_mask_sample_point_idxs =
      toFPSPointIdxs(in_mask_sh_points, in_mask_sample_polar_idxs,
                     sample_point_scale, anchor_num);
  std::cout << "toFPSPointIdxs : " << timer.now() << std::endl;

  const torch::Tensor in_mask_sample_points =
      in_mask_sh_points.index({fps_in_mask_sample_point_idxs});

  const torch::Tensor in_mask_sample_point_idxs =
      in_mask_sample_polar_idxs.index({fps_in_mask_sample_point_idxs});

  timer.reset();
  const torch::Tensor mask_boundary_sample_points =
      toSamplePoints(mask_degree_max, sh_degree_max, sh_params, rotate_vectors,
                     positions, mask_boundary_phis, mask_boundary_thetas,
                     mask_boundary_phi_idxs, use_inv);
  std::cout << "toSamplePoints : " << timer.now() << std::endl;

  const std::vector<torch::Tensor> sample_points_with_idxs(
      {mask_boundary_sample_points, in_mask_sample_points,
       in_mask_sample_point_idxs});

  return sample_points_with_idxs;
}
